experimental/beacon_sim/belief_road_map_planner.cc

```diff
@@ -1,14 +1,52 @@
 
 #include "experimental/beacon_sim/belief_road_map_planner.hh"
 
-#include <iostream>
 #include <random>
+#include <unordered_map>
 
+#include "Eigen/Core"
+#include "common/liegroups/se2.hh"
+#include "experimental/beacon_sim/ekf_slam.hh"
 #include "experimental/beacon_sim/generate_observations.hh"
 #include "experimental/beacon_sim/robot.hh"
 
 namespace robot::experimental::beacon_sim {
 namespace {
+using BeliefTransformMatrix =
+    Eigen::Matrix<double, 2 * liegroups::SE2::DoF, 2 * liegroups::SE2::DoF>;
+
+struct DirectedEdge {
+    int source;
+    int destination;
+    double initial_heading_in_local;
+
+    bool operator==(const DirectedEdge &other) const {
+        return source == other.source && destination == other.destination &&
+               initial_heading_in_local == other.initial_heading_in_local;
+    }
+};
+
+struct DirectedEdgeHash {
+    size_t operator()(const DirectedEdge &edge) const {
+        std::hash<int> int_hasher;
+        std::hash<double> double_hasher;
+        return (int_hasher(edge.source) << 3) ^ int_hasher(edge.destination) ^
```

Comment:
can you explain the hashing strategy?

Reply:
Sure thing! There are predefined hash functions for basic types (int, double) and I want to combine them to form a hash for this type. When combining the hash for the source and destination members, I shift one so that the hash isn't symmetric.